### ISSUE TITLE
[promise-function-async] Should not report for returned union types containing non-promise

### ISSUE THREAD

**Issue by sindresorhus** [typescript-eslint/typescript-eslint#1005]
```js
{
	'@typescript-eslint/promise-function-async': [
		'error',
		{
			allowAny: true
		}
	]
}
```

```ts
type Return = NodeJS.ReadableStream | Promise;

const promiseOrStream = (options: Options): Return => options.stream ? asStream(options) : asPromise(options);
```

**Expected Result**

I expected it to not report as adding an `async` keyword there will not work.

**Actual Result**

It reports.

**Additional Info**

It should only report when the return value is a `Promise`.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.3.1` |
| `@typescript-eslint/parser`        | `2.3.1` |
| `TypeScript`                       | `3.6.3` |
| `ESLint`                           | `6.4.0` |
| `node`                             | `12.10.0` |
| `npm`                              | `6.11.3` |

**Comment by a-tarasyuk** [comment]
@sindresorhus In order to reproduce the bug with the provided example `"allowAny"` should equal to `false`?
